### PR TITLE
[post] open-source for extension

### PR DESCRIPTION
**Pull request by TomasVotruba**
Could I ask you for post-review please? @martinahynkova

### CODE REVIEW

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+Magic? No, <strong>just patterns</strong>! Today we look on 3 of them I found and fond in Symfony and Laravel world.
+'''
+lang: en
+---
+
+## How is Open-Source Different from Private Code
+
+There is a big mind-shift from closed-source to open-source. To make it really work, you need to move from *my ego first* to *other people's feelings first*.
+
+It is like building Matrix open to everybody. **You have to predict future and unexpected use cases**. Your code have to be **extendable without making any changes to it**.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: možná..... changes "in" it

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+### "Opened for Extension, Closed for Modification"
+
+Now, I should refer to [**Open/closed principle** on Wikipedia](https://en.wikipedia.org/wiki/Open/closed_principle), which is the worst way to explain it.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: proč should? co takhle "Now I can...."

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+Instead, **I took a time to find simple example** (pro tip: Google with "simple") and actually found one - [go check it](https://github.com/wataridori/solid-php-example/blob/b84657cb736f86dda1453061d15df01f260e5140/2-open-closed-principle.php#L20-L32), it clearly show wrong approach.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: ....it clearly show**s**

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+}
+
+class ComputerEntrance extends BoothCallEntrance
+{
+
+}
+```
+
+**always [mark you classes final](https://ocramius.github.io/blog/when-to-declare-classes-final/)**. There is event [sniff for that](https://github.com/Symplify/CodingStandard/blob/master/src/Sniffs/Classes/FinalInterfaceSniff.php). Use it.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: asi jsi myslel?  "mark you**r** classes final"

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+}
+
+final ComputerEntrance implements MatrixEntranceInterface
+{
+
+}
+```
+
+Programmers won't have to think about raping your classes in the night and they just the interface you provide.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: "they just the interface you provide" nedává mi smysl, nevím jak jsi myslel?

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+<div class="text-center">
+    <img src="/../../../../assets/images/posts/2017/extendable-open-source/booth.png" class="thumbnail">
+</div>
+
+This approach is implemented in PHP under name of EventDispatcher. While working with Symfony, **events gave me very similar feeling of freedom** - [in docs](http://symfony.com/doc/current/reference/events.html#kernel-events) as well in small book [A Year with Symfony](https://leanpub.com/a-year-with-symfony).
+
+Do you want simple example of such listening script? [Check this tested post](https://pehapkari.cz/blog/2016/12/05/symfony-event-dispatcher/) with all code snippets you need.
+
+### While on Event, Listen Carefuly
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: carefully

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    public function runOnPing($location)
+    {
+        $this->smsSender->sendABoothAlert($location);
+    }
+}
+```
+
+### Why is this Useful?
+
+- You can introduce entriy point via event.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: entry

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+- You can **also pass metadata**, like location. Those data **can be open to change**, but don't have to be.
+
+It might be confusing while using at first, but after few weeks I get used to it. Trust me, it's the best.
+
+
+## 3. Like Collecting stamps, just on Steroids
+
+
+This is most powerfol and less know architecture pattern.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: powerfull

Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: known?

File: source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+All services of `EventSubscriber` type are collected by EventDispatcher.
+
+
+### You Probably Already Use It
+
+- Console Commands → Console Application
+- [Security Voters](http://symfony.com/doc/current/security/voters.html) → Access Decision Manager
+
+
+As for tags - [they often promote bad practise of duplicated information](https://www.tomasvotruba.cz/blog/2017/02/12/drop-all-service-tags-in-your-nette-and-symfony-applications/#bare-tagging-is-duplicated-information). Don't use it if don't have to.
```
Comment (martinahynkova) [inline source/_posts/2017/2017-04-14-3-symfony-and-laravel-patterns-that-make-code-easy-to-extends-without-modification.md]: .....if **you** don´t have to.

**Comment by TomasVotruba** [comment]
Thank @martinahynkova :heart:

**Pull request merged**